content/news/2020-08-28-gsoc-ends-2020.md

```diff
@@ -0,0 +1,56 @@
++++
+title = "Google Summer of Code Wrap-up"
+date = 2020-08-28
+extra.author = "Devin Ragotzy"
++++
+
+Wow the summer has flown by, it feels like just yesterday I was learning how to rebase and what exactly it is ruma does. I exaggerate slightly, but it is a big library with lots of public API surface. I have learned more in the last few months than in two years of school. I have been able to observe and participate in a project with a community growing around it, been apart of discussions about design and best practices, given and received numerous code reviews as well as learned the process of addressing the feedback, and working from a specification. In short this has been an amazing opportunity to gain experience in all the things that are hard to obtain in a classroom.
+
+My project goal was to improve the existing macros in ruma-events-macros and ruma-api-macros. It became clear early on that this would include some major API changes and that improving the macros as they were was pointless without also moving to the new API design. While improving the durability and readability of the macro code I also rewrote entire sections to accommodate the new design.
+
+A quick overview of the Matrix protocol for reference: a client sends events to a server and that server distributes those events out to other clients and other servers (the server case is known as federation). Ruma is the library that represents these events. The events are grouped by kind Message, State, Ephemeral, ToDevice and Basic which ruma represents as generic structs (`StateEvent<C>`). Each event kind needs to be able to hold many different content types, for State events there is room creation, room name, and membership events to name a few. Using the macros enums are generated to represent all State event possibilities, so a variant for membership, room name, etc. These types are to support the core API Request and Response types for each endpoint that is defined by the matrix specification.
```

Comment:
> a client sends events

Technically wrong, since event content != event, but I'm not sure how to improve this sentence in a nice way.

> Ruma is the library that represents these events.

We do a bit more than that 😛
Did you mean to refer to just ruma-events?

> The events are grouped by […]

The spec isn't actually very explicit about this. I think "basic event" is entirely our own terminology. Not sure whether / how to best clarify though.

> State events

Should be either `state events` or `State Events` (I think I prefer the former).

> These types are to support

"exist to support"? Sth. else?

```suggestion
A quick overview of the Matrix protocol for reference: a client sends events to a server and that server distributes those events out to other clients and other servers (the server case is known as federation). Ruma is the library that represents these events. The events are grouped by kind Message, State, Ephemeral, ToDevice and Basic which Ruma represents as generic structs (`StateEvent<C>`). Each event kind needs to be able to hold many different content types, for State events there is room creation, room name, and membership events to name a few. Using the macros, enums are generated to represent all State event possibilities, so a variant for membership, room name, etc. These types are to support the core API request and response types for each endpoint that is defined by the matrix specification.
```